frontend/catalyst/compiler.py

```diff
@@ -249,6 +238,13 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
     ],
 )
 
+BUFFERIZATION_ASYNC_PASS = (
+    "BufferizationPass",
+    [
+        s.replace("}", " copy-before-write}") if s.startswith("one-shot-bufferize") else s
+        for s in BUFFERIZATION_PASS[1]
+    ],
+)
```

Comment:
This one is quite tricky. For the async tests like below, if we do not add this flag, the result will be 0 (wrong).

```python
import pennylane as qml
from catalyst import  qjit,

def test_gradient(inp, diff_methods, backend):
    """Parameter shift and finite diff generate multiple QNode that are run async."""

    def f(x):
        qml.RX(x * 2, wires=0)
        return qml.expval(qml.PauliY(0))

    @qjit(async_qnodes=True, keep_intermediate=True)
    def compiled(x: float):
        g = qml.qnode(qml.device(backend, wires=1), diff_method=diff_methods[0])(f)
        h = grad(g, method=diff_methods[1])
        return h(x)

    def interpreted(x):
        device = qml.device("default.qubit", wires=1)
        g = qml.QNode(f, device, diff_method="backprop")
        h = qml.grad(g, argnum=0)
        return h(x)

    return compiled(inp)

test_gradient((1.0), ("finite-diff", "fd"), "lightning.qubit")
``` 

Bufferized IR without the flag (incorrect result):
```mlir
    %2 = call @f(%arg0) : (memref<f64>) -> memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%arg0, %1 : memref<f64>, memref<f64>) outs(%arg0 : memref<f64>) {
    ^bb0(%in: f64, %in_0: f64, %out: f64):
      %9 = arith.addf %in, %in_0 : f64
      linalg.yield %9 : f64
    }
    %3 = call @f(%arg0) : (memref<f64>) -> memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%3, %2 : memref<f64>, memref<f64>) outs(%3 : memref<f64>) {
    ^bb0(%in: f64, %in_0: f64, %out: f64):
      %9 = arith.subf %in, %in_0 : f64
      linalg.yield %9 : f64
    }
```

Bufferized IR with the flag (correct result):
```mlir
    %alloc = memref.alloc() {alignment = 64 : i64} : memref<f64>
    memref.copy %arg0, %alloc : memref<f64> to memref<f64>
    %2 = call @f(%alloc) : (memref<f64>) -> memref<f64>
    %alloc_0 = memref.alloc() {alignment = 64 : i64} : memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%arg0, %1 : memref<f64>, memref<f64>) outs(%alloc_0 : memref<f64>) {
    ^bb0(%in: f64, %in_4: f64, %out: f64):
      %9 = arith.addf %in, %in_4 : f64
      linalg.yield %9 : f64
    }
    %alloc_1 = memref.alloc() {alignment = 64 : i64} : memref<f64>
    memref.copy %alloc_0, %alloc_1 : memref<f64> to memref<f64>
    memref.dealloc %alloc_0 : memref<f64>
    %3 = call @f(%alloc_1) : (memref<f64>) -> memref<f64>
    %alloc_2 = memref.alloc() {alignment = 64 : i64} : memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%3, %2 : memref<f64>, memref<f64>) outs(%alloc_2 : memref<f64>) {
    ^bb0(%in: f64, %in_4: f64, %out: f64):
      %9 = arith.subf %in, %in_4 : f64
      linalg.yield %9 : f64
    }
```
I think `%arg0` is somehow changed and makes the results incorrect. `copy-before-write` creates copies of `%arg0` could protect it from modified.

Reply:
I would like if possible to merge #1194 before this bufferization. Then you could just use:

```python
  "one-shot-bufferize{bufferize-function-boundaries allow-return-allocs-from-loops}" if not options.async_qnodes else ...
```

Reply:
I think this flag essentially disables bufferization. I am not opposed to disabling bufferization for the async special case, but we will have to fix that in the future.

Reply:
It does disable the analysis for the buffererization, meaning we fall back to worst case (bufferization itself obv still takes place since we can't compile without having buffers for the tensors). What I'm wondering, is this worse than before, because I think the old bufferization was also pretty poor performance? 🤔 

We briefly talked about the example Tzung-Han gave, I think it's easy to understand where it goes wrong:
```
    %2 = call @f(%arg0) : (memref<f64>) -> memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%arg0, %1 : memref<f64>, memref<f64>) outs(%arg0 : memref<f64>) {
```
if `f` is a qnode and it executes asynchronously, and the next instruction writes to `%arg0` (since it appears in the `outs` parameter of the linalg op), it could mess up the results.

At least for this sort of code, I would be surprised if the old bufferization did anything in-place. But I guess the actual question is how do we let the bufferization know about this sort of thing?

Reply:
Yes, previously the values passed in as arguments were the responsibility of the caller to be deallocated. As a result `%arg0` would never be deallocated by `jit_compiled`. See below for a snippet of the IR after the bufferization:

```mlir
  func.func public @jit_compiled(%arg0: memref<f64>) -> memref<f64> attributes {llvm.copy_memref, llvm.emit_c_interface} {
    %0 = llvm.mlir.constant(3735928559 : index) : i64
    %1 = memref.get_global @__constant_xf64 : memref<f64>
    %2 = call @f_0(%arg0) : (memref<f64>) -> memref<f64>
    %alloc = memref.alloc() {alignment = 64 : i64} : memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%arg0, %1 : memref<f64>, memref<f64>) outs(%alloc : memref<f64>) {
    ^bb0(%in: f64, %in_2: f64, %out: f64):
      %9 = arith.addf %in, %in_2 : f64
      linalg.yield %9 : f64
    }
    %3 = call @f_0(%alloc) : (memref<f64>) -> memref<f64>
    %alloc_0 = memref.alloc() {alignment = 64 : i64} : memref<f64>
    linalg.generic {indexing_maps = [affine_map<() -> ()>, affine_map<() -> ()>, affine_map<() -> ()>], iterator_types = []} ins(%3, %2 : memref<f64>, memref<f64>) outs(%alloc_0 : memref<f64>) {
    ^bb0(%in: f64, %in_2: f64, %out: f64):
      %9 = arith.subf %in, %in_2 : f64
      linalg.yield %9 : f64
    }
```

I do feel that a copy is a regression though. The IR in its current state has no copy but the flag inserts this:

```
    %alloc_1 = memref.alloc() {alignment = 64 : i64} : memref<f64>
    memref.copy %alloc_0, %alloc_1 : memref<f64> to memref<f64>
    memref.dealloc %alloc_0 : memref<f64>
```

I do wonder if any of the optimizations at the buffer level helps.

Reply:
Got it! Yeah that's not great, although at least it only applies to async. Still not great though :/

Reply:
I think to fix this we need some changes in how we lower to qnodes to async. Still thinking :thinking: